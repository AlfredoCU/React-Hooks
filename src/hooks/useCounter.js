import { useState } from "react";

export const useCounter = (count = 0) => {
  const [counter, setCounter] = useState(count);

  const increment = (factor = 1) => {
    setCounter(counter + factor);
  };

  const reset = () => {
    setCounter(count);
  };

  const decrement = (factor = 1, limit = 1) => {
    if (counter > limit) {
      setCounter(counter - factor);
    }
  };

  //* Retornamos un obj, pero tambiÃ©n puede ser un array, depende de las necesidades.
  return {
    counter,
    increment,
    reset,
    decrement,
  };
};
