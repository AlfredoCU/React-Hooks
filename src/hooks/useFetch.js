import { quotesBb } from "../helpers/quotesBb";
import { useEffect, useState, useRef } from "react";

export const useFetch = (id) => {
  const [state, setstate] = useState({ data: null, loading: true });
  const isMounted = useRef(true); // Montado.
  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    if (isMounted.current) {
      quotesBb(id)
        .then((data) => setstate({ data, loading: false }))
        .catch((error) => console.error("USE_FETCH_ERROR", error));
    } else {
      console.log("isUnmountend"); // Desmontado.
    }
  }, [id]);

  return state;
};
