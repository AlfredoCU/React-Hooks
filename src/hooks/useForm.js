import { useState } from "react";

export const useForm = (quantity = {}) => {
  const [inputs, setInputs] = useState(quantity);

  const reset = () => {
    setInputs(quantity);
  };

  const handleInputChange = ({ target }) => {
    setInputs({
      ...inputs,
      [target.name]: target.value,
    });
  };

  return [inputs, handleInputChange, reset];
};

// console.log({
//! Mantiene el estado anterio, sino se agrega -> y escribes en name y despues en email, se pierde el estado anterio de name.
//   ...inputs,
//   [target.name]: target.value,
// });
