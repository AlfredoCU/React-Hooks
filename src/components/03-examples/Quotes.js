import React from "react";
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";

const Quotes = () => {
  const { counter, increment, reset, decrement } = useCounter(1);
  const { data, loading } = useFetch(counter);
  // data: [...] -> Estrictamente debe de ser un true para que pase a la segunda parete.
  const { author, quote } = !!data && data[0]; // data: null -> no entra a la segunda parte. ya que seria forzozamente false.

  return (
    <div>
      <h4>API BREAKING BAD</h4>
      {loading ? (
        <div className="alert alert-info text-center">Cargando...</div>
      ) : (
        <div>
          <p>{counter}</p>
          <p className="blockquote">{quote}</p>
          <em className="blockquote-footer mb-2">{author}</em>
          <div className="mb-2">
            {/* Es necesario esta sintaxis para no obtener como argumento el evento Dom. */}
            <button
              className="btn btn-primary mr-2"
              onClick={() => decrement()}
            >
              Atras
            </button>
            {/* No es necesario ya que no recibe ningun argumeto. */}
            <button className="btn btn-primary mr-2" onClick={reset}>
              Reset
            </button>
            {/* Podemos enviar un argumento de incremento y uno de limite, tambien en decremente. */}
            <button
              className="btn btn-primary mr-2"
              onClick={() => increment(1)}
            >
              Adelante
            </button>
          </div>
        </div>
      )}
    </div>
  );
};

export default Quotes;

// {
//   data &&
//     data.map((item, key) => (
//       <li key={`${item.quote_id} - quote - ${key}`}>
//         {item.quote} - {item.author} - {item.series}
//       </li>
//     ));
// }
