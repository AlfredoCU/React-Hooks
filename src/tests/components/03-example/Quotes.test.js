import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";

import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";
import Quotes from "../../../components/03-examples/Quotes";

//! Fingir la data.
jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");

describe("Prueba del componente Quotes", () => {
  //! TambiÃ©n se puede mandar a llamar otros hooks con data falsa.
  useCounter.mockReturnValue({
    counter: 1,
    increment: () => {},
    reset: () => {},
    decrement: () => {},
  });

  test("Foto del componente Cargado...", () => {
    useFetch.mockReturnValue({ data: null, loading: true }); //! Data inicial.
    const wrapper = shallow(<Quotes />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert").exists()).toBe(true);
  });

  test("Foto del componente con data", () => {
    const info = {
      author: "Alfredo",
      quote: "Hello World!",
    };

    useFetch.mockReturnValue({ data: [info], loading: false }); //! Data falsa.
    const wrapper = shallow(<Quotes />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert").exists()).toBe(false);
    expect(wrapper.find("em").text().trim()).toBe("Alfredo");
    expect(wrapper.find("button").exists()).toBe(true);
  });
});

/*
  Cuando tengas que evaluar las funciones de un componente, con la biblioteca jest es suficiente
  utilizando simulate().
  Cuando tengas que evaluar las funciones de un custom hook, con la biblioteca render-hooks
  utiliza act(). 
*/
