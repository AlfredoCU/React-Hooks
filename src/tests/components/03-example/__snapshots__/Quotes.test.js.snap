// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba del componente Quotes Foto del componente Cargado... 1`] = `
<div>
  <h4>
    API BREAKING BAD
  </h4>
  <div
    className="alert alert-info text-center"
  >
    Cargando...
  </div>
</div>
`;

exports[`Prueba del componente Quotes Foto del componente con data 1`] = `
<div>
  <h4>
    API BREAKING BAD
  </h4>
  <div>
    <p>
      1
    </p>
    <p
      className="blockquote"
    >
      Hello World!
    </p>
    <em
      className="blockquote-footer mb-2"
    >
      Alfredo
    </em>
    <div
      className="mb-2"
    >
      <button
        className="btn btn-primary mr-2"
        onClick={[Function]}
      >
        Atras
      </button>
      <button
        className="btn btn-primary mr-2"
        onClick={[Function]}
      >
        Reset
      </button>
      <button
        className="btn btn-primary mr-2"
        onClick={[Function]}
      >
        Adelante
      </button>
    </div>
  </div>
</div>
`;
